### PR TITLE
Fix RecurrentGemma device_map

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ?
This PR makes gemma compatible with multi-gpu device_map. To try out: 
```python 
from transformers import AutoTokenizer, AutoModelForCausalLM

tokenizer = AutoTokenizer.from_pretrained("google/recurrentgemma-2b-it")
model = AutoModelForCausalLM.from_pretrained(
    "google/recurrentgemma-2b-it", device_map="auto"
)
input_text = "Write me a poem about Machine Learning."
input_ids = tokenizer(input_text, return_tensors="pt").to("cuda")

outputs = model.generate(**input_ids,use_cache=True)
print(tokenizer.decode(outputs[0]))
```
I get the same output in the single gpu or multi gpu setup.

### CODE REVIEW

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 1]
```diff
@@ -252,7 +252,7 @@ def _update_cache(self, key_states, value_states, **cache_kwargs):
             to_shift = cache_position >= self.config.attention_window_size - 1
             indices = (slicing + to_shift[-1].int() - 1) % self.config.attention_window_size
 
-            k_out, v_out = self.key_states, self.value_states
+            k_out, v_out = self.key_states.to(key_states.device), self.value_states.to(value_states.device)
```
Comment (SunMarc) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:255]: Due to _setup_cache, self.key_states and self.value_states are initialized on the device of the hidden state that we pass to the model in generate (e.g. cuda:0). However, this layer might not be on the same device of the hidden state if we use multi-gpu. Hence, we need to make sure that `self.key_states` is on the same device as `key_states`. Same for `value_states`.

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 2]
```diff
@@ -376,7 +376,9 @@ def _rnn_scan(
                 return hidden_states, hidden_states[:, 0].type(acc_dtype)
 
             else:
-                contextualized_states = recurrent_gate.type(acc_dtype) * recurrent_states[:, None]
+                contextualized_states = recurrent_gate.type(acc_dtype) * recurrent_states[:, None].to(
+                    recurrent_gate.device
+                )
```
Comment (SunMarc) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:381]: Same issue with `recurrent_gate` which is initialized in `_setup_cache`.

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 3]
```diff
@@ -387,7 +389,7 @@ def _rnn_scan(
 
             contextualized_states = torch.zeros_like(hidden_states)
             for t in range(hidden_states.shape[1]):
-                recurrent_states = recurrent_gate[:, t].type(acc_dtype) * recurrent_states
+                recurrent_states = recurrent_gate[:, t].type(acc_dtype) * recurrent_states.to(recurrent_gate.device)
```
Comment (SunMarc) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:392]: Here also !

File: src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py [hunk 4]
```diff
@@ -658,7 +660,9 @@ def __init__(self, config: RecurrentGemmaConfig):
         self.final_norm = RecurrentGemmaRMSNorm(config.hidden_size, eps=config.rms_norm_eps)
         self.gradient_checkpointing = False
 
-        self.register_buffer("normalizer", torch.tensor(self.config.hidden_size**0.5, dtype=torch.bfloat16))
+        self.register_buffer(
+            "normalizer", torch.tensor(self.config.hidden_size**0.5, dtype=torch.bfloat16), persistent=False
+        )
```
Comment (SunMarc) [inline src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:665]: We don't need this to be persistant. This fixes an issue that we get with accelerate too.

Comment (ArthurZucker) [reply src/transformers/models/recurrent_gemma/modeling_recurrent_gemma.py:665]: Good catch

**Review by ArthurZucker — COMMENTED** [review]
Thanks, the device thing could be fixed by placing them on the same device as self.key_states? rather than the device passed? 
Also tad bit scared of the slow down of doing it there? But LGTM otherwise

**Comment by SunMarc** [comment]
> Thanks, the device thing could be fixed by placing them on the same device as self.key_states? rather than the device passed? Also tad bit scared of the slow down of doing it there? But LGTM otherwise

I think it will slow down if why place them on the same device as self.key_states for example. Let's say `self.key_states` is initialized on cuda:0 and we have 2 gpus. The problem is that the computed key_states can be on cuda:0 or cuda:1 depending on where the layer is. Hence, it is better to move` self.key_states` to the device of `key_states` to limit data transfert between gpus. Otherwise, we will need to move the data each time we have a layer in cuda:1.

**Review by ArthurZucker — APPROVED** [review]
Thanks

**Pull request merged**